File: runner/src/integration-tests/java/com/daylifecraft/minigames/games/RoundSearchProviderTest.java
package com.daylifecraft.minigames.games;

import com.daylifecraft.minigames.UtilsForTesting;
import com.daylifecraft.minigames.minigames.PlayerMiniGameManager;
import com.daylifecraft.minigames.minigames.search.IRoundSearchProvider;
import net.minestom.server.entity.Player;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;
import org.mockito.Mockito;

class RoundSearchProviderTest {

  private static final IRoundSearchProvider roundSearchProvider =
    Mockito.spy(IRoundSearchProvider.class);

  private static final Player fakePlayer = UtilsForTesting.initFakePlayer("GProviderT");

  @BeforeAll
  static void setup() throws InterruptedException {
    UtilsForTesting.waitUntilPlayerJoin(fakePlayer);
  }

  @Test
  void testDoesPlayerCannotBePreparedWhenLocked() {
    try (MockedStatic<PlayerMiniGameManager> playerMiniGameManager =
           Mockito.mockStatic(PlayerMiniGameManager.class)) {
      playerMiniGameManager
        .when(() -> PlayerMiniGameManager.isPlayerLocked(fakePlayer.getUuid()))
        .thenReturn(true);

      Assertions.assertFalse(
        roundSearchProvider.canBePrepared(fakePlayer),
        "Expected that locked player cannot be prepared");
    }
  }

  @Test
  void testDoesPlayerCanBePrepared() {
    try (MockedStatic<PlayerMiniGameManager> playerMiniGameManager =
           Mockito.mockStatic(PlayerMiniGameManager.class)) {
      playerMiniGameManager
        .when(() -> PlayerMiniGameManager.isPlayerLocked(fakePlayer.getUuid()))
        .thenReturn(false);

      Assertions.assertTrue(
        roundSearchProvider.canBePrepared(fakePlayer),
        "Expected that does not locked player can be prepared");
    }
  }

  @AfterAll
  static void kickPlayer() {
    fakePlayer.kick("");
  }
}

File: runner/src/integration-tests/java/com/daylifecraft/minigames/games/RoundSearchProviderTest.kt
package com.daylifecraft.minigames.games

import com.daylifecraft.minigames.UtilsForTesting
import com.daylifecraft.minigames.minigames.PlayerMiniGameManager
import com.daylifecraft.minigames.minigames.search.IRoundSearchProvider
import io.mockk.every
import io.mockk.mockkObject
import io.mockk.spyk
import net.minestom.server.entity.Player
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

internal class RoundSearchProviderTest {
  @Test
  fun testDoesPlayerCannotBePreparedWhenLocked() {
    mockkObject(PlayerMiniGameManager) {
      every { PlayerMiniGameManager.isPlayerLocked(fakePlayer.uuid) } returns true

      assertFalse(
        roundSearchProvider.canBePrepared(fakePlayer),
        message = "Expected that locked player cannot be prepared",
      )
    }
  }

  @Test
  fun testDoesPlayerCanBePrepared() {
    mockkObject(PlayerMiniGameManager) {
      every { PlayerMiniGameManager.isPlayerLocked(fakePlayer.uuid) } returns false

      assertTrue(
        roundSearchProvider.canBePrepared(fakePlayer),
        message = "Expected that does not locked player can be prepared",
      )
    }
  }

  companion object {
    private val roundSearchProvider = spyk<IRoundSearchProvider>()

    private val fakePlayer: Player = UtilsForTesting.initFakePlayer("GProviderT")

    @BeforeAll
    @Throws(InterruptedException::class)
    @JvmStatic
    fun setup() {
      UtilsForTesting.waitUntilPlayerJoin(fakePlayer)
    }

    @AfterAll
    @JvmStatic
    fun kickPlayer() {
      fakePlayer.kick("")
    }
  }
}
